redhat-appstudio/service-provider-integration-operator
# Token matching for Quay

### What does this PR do?
Initial implementation of token matching for Quay. This should support both OAuth tokens and robot tokens.

The only limitation is that we explicitly DON'T support `user:read` and `user:admin` scopes (and related
SPI-specific permissions) for Quay, because there is no way of telling which user the token should be
provided for, given just a repository URL. If these permissions are required, we can think about how to
implement this at some later point in time.

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-89

### How to test this PR?
1. Install SPI.
   (replace `deploy_minikube` with `deploy_openshift` below if you're testing on OpenShift)
   ```
   make install deploy_minikube SPIO_IMG=quay.io/lkrejci/service-provider-integration-operator:svpi-89
   ```
1. Edit the SPI configuration. Enter some "sharedSecret" and enter the client ID and client Secret of your Quay OAuth application. Also CHANGE THE `baseUrl` TO BE A `https` URL:
   ```
   ./hack/edit-spi-config.sh
   ```
1. Restart the deployments
   ```
   kubectl -n spi-system scale deployment spi-controller-manager spi-oauth-service --replicas=0
   kubectl -n spi-system scale deployment spi-controller-manager spi-oauth-service --replicas=1
   ```
1. Create a binding for pulling the repository (the below repository is under my user and is public. Any token should be able to read it):
    ```
    kubectl apply -f - <<END
    apiVersion: appstudio.redhat.com/v1beta1
    kind: SPIAccessTokenBinding
    metadata:
      name: binding-pull-repository
    spec:
      permissions:
        required:
        - type: r
          area: repository
      repoUrl: quay.io/lkrejci/alloc
      secret:
        name: token-pull-repository
        type: kubernetes.io/basic-auth
    END
    ```
1. Create binding for reading the repository metadata (the below is a binding for a private repo. If you don't have access to `redhat_emp1` organization in Quay, replace the repository with some of your own):
    ```
    kubectl apply -f - <<END
    apiVersion: appstudio.redhat.com/v1beta1
    kind: SPIAccessTokenBinding
    metadata:
      name: binding-read-repository
    spec:
      permissions:
        required:
        - type: r
          area: repositoryMetadata
      repoUrl: https://quay.io/redhat_emp1/lkrejci_spi_private_repo_test
      secret:
        name: token-read-repository
        type: kubernetes.io/basic-auth
    END
    ```
1. Create a binding for pulling and pushing the repository (the below repository is under my user and is public. Replace it with a repository that you can push into):
    ```
    kubectl apply -f - <<END
    apiVersion: appstudio.redhat.com/v1beta1
    kind: SPIAccessTokenBinding
    metadata:
      name: binding-push-repository
    spec:
      permissions:
        required:
        - type: rw
          area: repository
      repoUrl: https://quay.io/lkrejci/alloc
      secret:
        name: token-push-repository
        type: kubernetes.io/basic-auth
    END
    ```
1. Create a binding for reading and writing repository metadata (again, replace the repository if needed):
    ```
    kubectl apply -f - <<END
    apiVersion: appstudio.redhat.com/v1beta1
    kind: SPIAccessTokenBinding
    metadata:
      name: binding-write-repository
    spec:
      permissions:
        required:
        - type: rw
          area: repositoryMetadata
      repoUrl: https://quay.io/redhat_emp1/lkrejci_spi_private_repo_test
      secret:
        name: token-write-repository
        type: kubernetes.io/basic-auth
    END
    ```
1. Give the default service account the ability to work with SPI objects:
    ```
    kubectl apply -f hack/give-default-sa-perms-for-accesstokens.yaml
    ```
1. Create a robot account under your user and give it write access to the repository mentioned in the `binding-push-repository` and `binding-pull-repository`.
1. Upload the robot account to the spi access token backing the `binding-push-repository`:
    ```
    ROBOT_TOKEN_NAME=<...>
    ROBOT_TOKEN_PASSWORD=<...>
    curl -k -v -X POST -H "Content-Type: application/json" -H "Authorization: Bearer $(hack/get-default-sa-token.sh)" -d'{"username": "'$ROBOT_TOKEN_NAME'", "access_token": "'$ROBOT_TOKEN_PASSWORD'"}' "http://spi.$(minikube ip).nip.io/token/default/$(kubectl get spiaccesstokenbinding binding-push-repository -o yaml | yq '.status.linkedAccessTokenName')"
    ```
    Check that both `binding-push-repository` and `binding-pull-repository` bindings are now injected.
1. In OAuth service repository, open the simple OAuth UI:
    ```
    cd ......./service-provider-integration-oauth
    firefox hack/oauth-ui.html
    ```
1. Go through the OAuth flow for the `binding-read-repository` binding.
    1. Take the OAuth URL for the `binding-read-repository` using this command:
        ```
        kubectl get spiaccesstokenbinding binding-read-repository -o yaml | yq '.status.oAuthUrl'
        ```
        and copy it to the "OAuth Url" field on the page.
    1. Copy the token of the default SA (get it by running `hack/get-default-sa-token.sh`) to the "K8s token" field on the page
    1. Click on the "Login" button
    1. Go back in browser history
    1. Click on the "Initiate OAuth" button
    1. Grant the permission in the Quay UI opened in the a new tab
1. Check that the `binding-read-repository` binding is now injected and that `binding-write-repository` is still awaiting data.
1. Go through the OAuth flow for `binding-write-repository` binding. Repeat the steps for the OAuth flow above
1. Check that `binding-write-repository` is now also injected
